tpu-client-next/src/node_address_service.rs

```diff
@@ -0,0 +1,1549 @@
+
+        // If we are close to the end of slot, increment the estimated current slot
+        if timestamp() - start_time >= estimated_slot_duration.as_millis() as u64 {
+            estimated_current_slot = estimated_current_slot.saturating_add(1);
+        }
+
+        let (last_slot, last_slot_in_epoch, slots_in_epoch) = self.slot_info();
+
+        debug!(
+            "Estimated current slot: {}, last slot: {}, last slot in epoch: {}, slots in epoch: {}",
+            estimated_current_slot, last_slot, last_slot_in_epoch, slots_in_epoch
+        );
+        // If we're crossing into a new epoch, fetch the updated epoch schedule.
+        if estimated_current_slot > last_slot_in_epoch {
+            debug!(
+                "Crossing into a new epoch, fetching updated epoch schedule. \
+                 Last slot in epoch: {}, estimated current slot: {}",
+                last_slot_in_epoch, estimated_current_slot
+            );
+            match rpc_client.get_epoch_schedule().await {
+                Ok(epoch_schedule) => {
+                    let epoch = epoch_schedule.get_epoch(estimated_current_slot);
+                    self.slots_in_epoch = epoch_schedule.get_slots_in_epoch(epoch);
+                    self.last_slot_in_epoch = epoch_schedule.get_last_slot_in_epoch(epoch);
+                }
+                Err(err) => {
+                    warn!("Failed to fetch epoch schedule: {err}");
+                    result = Err(NodeAddressServiceError::RpcError(err.into()));
+                }
+            }
+        }
+
+        // If we are within the fanout range of the last slot in the cache,
+        // fetch more slot leaders. We pull down a big batch at at time to
+        // amortize the cost of the RPC call. We don't want to stall
+        // transactions on pulling this down so we fetch it proactively.
+        if estimated_current_slot >= last_slot.saturating_sub(MAX_FANOUT_SLOTS) {
+            let slot_leaders = rpc_client
+                .get_slot_leaders(
+                    estimated_current_slot,
+                    LeaderTpuCacheServiceState::fanout(slots_in_epoch),
+                )
+                .await;
+            match slot_leaders {
+                Ok(slot_leaders) => {
+                    self.first_slot = estimated_current_slot;
+                    self.leaders = slot_leaders;
+                }
+                Err(err) => {
+                    warn!(
+                        "Failed to fetch slot leaders (first_slot: \
+                         {}): {err}",
+                        estimated_current_slot
+                    );
+                    result = Err(NodeAddressServiceError::RpcError(err.into()));
+                }
+            }
+        }
+        result
+    }
+}
+
+// 48 chosen because it's unlikely that 12 leaders in a row will miss their slots
+const MAX_SLOT_SKIP_DISTANCE: u64 = 48;
+
+#[derive(Debug)]
+pub(crate) struct RecentLeaderSlots(VecDeque<Slot>);
```

Comment:
Probably, this code will be replaced with something that is able to predict the slot duration but not in the first pr